### DISCUSSION TITLE
TV-box TX95. Wi-Fi module  AP6330

### DISCUSSION THREAD

**Discussion by dmotkan** [devmfc/debian-on-amlogic#36] [Ideas]
Hi , devmfc .
TV-box T95W2.  Amlogic S905W2 .    Wi-Fi module  AP6330  insaide, under the lid , Broadcom bcm4330 .
[https://4pda.to/forum/index.php?showtopic=1076147&view=findpost&p=128251554](url)
Is it possible, with this information, to make Wi-Fi support in Armbian for this TV box?

**Comment by dmotkan** [comment]
In order not to multiply topics, I will ask an additional question here.
Problem with the latest build release. It is not possible to install additional packages on the SD card because the system reports that the disk is 709 MB in size and is full. However, the partition size is 7701 MB. I don't understand what the problem is. Maybe there is a quota set for /root somewhere? This was not the case in previous releases. I didn't find a quota in /etc/fstab.
Here is the console output.

`Up time:       2 min              Boot time:     10 s
CPU usage:     0%                 Memory usage:  8% of 1861M
CPU temp:      __°C               DDR temp:      __°C
Boot device:   /dev/mmcblk0p1     Usage /boot:   14% of 269M
Root FS:       /dev/mmcblk0p2     Usage / :      100% of 708M

tvbox:~#parted -l
Error: /dev/mmcblk1boot0: unrecognised disk label
Model: Generic SD/MMC Storage Card (sd/mmc)
Disk /dev/mmcblk1boot0: 4194kB
Sector size (logical/physical): 512B/512B
Partition Table: unknown
Disk Flags:

Model: SD SD (sd/mmc)
Disk /dev/mmcblk0: 7986MB
Sector size (logical/physical): 512B/512B
Partition Table: msdos
Disk Flags:

Number  Start   End     Size    Type     File system  Flags
 1      16.8MB  285MB   268MB   primary  fat16        lba
 2      285MB   7986MB  7701MB  primary  ext4

Model: Unknown (unknown)
Disk /dev/zram1: 157MB
Sector size (logical/physical): 4096B/4096B
Partition Table: loop
Disk Flags:

Number  Start  End    Size   File system  Flags
 1      0.00B  157MB  157MB  ext4

Error: /dev/mmcblk1boot1: unrecognised disk label
Model: Generic SD/MMC Storage Card (sd/mmc)
Disk /dev/mmcblk1boot1: 4194kB
Sector size (logical/physical): 512B/512B
Partition Table: unknown
Disk Flags:

Model: MMC DF4016 (sd/mmc)
Disk /dev/mmcblk1: 15.8GB
Sector size (logical/physical): 512B/512B
Partition Table: msdos
Disk Flags:

Number  Start   End     Size    Type     File system  Flags
 1      742MB   1279MB  537MB   primary               lba
 2      1279MB  15.8GB  14.5GB  primary


Model: Unknown (unknown)
Disk /dev/zram0: 786MB
Sector size (logical/physical): 4096B/4096B
Partition Table: loop
Disk Flags:

Number  Start  End    Size   File system     Flags
 1      0.00B  786MB  786MB  linux-swap(v1)`

**Reply by devmfc** [reply]
The rootfs filesystem did not expand properly during first boot. 
Maybe the first boot was interrupted by a reset or power cycle or crash?
Try to flash the sd card again and try again.

**Comment by dmotkan** [comment]
~#bash aml-install-to-emmc.sh

Current Host disk:  [ /dev/mmcblk0 ]
Target eMMC disk: [ /dev/mmcblk1 ]
Select partition table type [ MSDOS ]
This script will erase all existing partitions and its data on the disk!!
Are you sure y/N ? y
Ok, proceeding...
Start create partition table and filesystems on eMMC [ /dev/mmcblk1 ]...
Backup vendor u-boot
aml-install-to-emmc.sh: line 111: ./aml-part.sh: Permission denied
aml-install-to-emmc.sh: line 112: ./aml-part.sh: Permission denied
aml-install-to-emmc.sh: line 113: ./aml-part.sh: Permission denied
aml-install-to-emmc.sh: line 114: ./aml-part.sh: Permission denied
Error: Input/output error during write on /dev/mmcblk1
Error occured [1], create_partition_table:120, exitting

I am attaching the answer from the previous working version:
`#bash aml-part.sh
Found valid Amlogic Partition Table with 29 partitions
0 MiB --- offset: 0x0 B --- 0 B --- 0 s ---
name: bootloader
flags: 0
size: 4 MiB 0x400000 B
end: 4 MiB 0x400000 B
GAP 32 MiB
36 MiB --- offset: 0x2400000 B --- 37748736 B --- 73728 s ---
name: reserved
flags: 0
size: 64 MiB 0x4000000 B
end: 100 MiB 0x6400000 B
GAP 8 MiB
108 MiB --- offset: 0x6c00000 B --- 113246208 B --- 221184 s ---
name: cache
flags: 0
size: 0 MiB 0x0 B
end: 108 MiB 0x6c00000 B
GAP 8 MiB
116 MiB --- offset: 0x7400000 B --- 121634816 B --- 237568 s ---
name: env
flags: 0
size: 8 MiB 0x800000 B
end: 124 MiB 0x7c00000 B
GAP 8 MiB
132 MiB --- offset: 0x8400000 B --- 138412032 B --- 270336 s ---
name: frp
flags: 1
size: 2 MiB 0x200000 B
end: 134 MiB 0x8600000 B
GAP 8 MiB
142 MiB --- offset: 0x8e00000 B --- 148897792 B --- 290816 s ---
name: factory
flags: 9
size: 8 MiB 0x800000 B
end: 150 MiB 0x9600000 B
GAP 8 MiB
158 MiB --- offset: 0x9e00000 B --- 165675008 B --- 323584 s ---
name: vendor_boot_a
flags: 1
size: 24 MiB 0x1800000 B
end: 182 MiB 0xb600000 B
GAP 8 MiB
190 MiB --- offset: 0xbe00000 B --- 199229440 B --- 389120 s ---
name: vendor_boot_b
flags: 1
size: 24 MiB 0x1800000 B
end: 214 MiB 0xd600000 B
GAP 8 MiB
222 MiB --- offset: 0xde00000 B --- 232783872 B --- 454656 s ---
name: tee
flags: 1
size: 32 MiB 0x2000000 B
end: 254 MiB 0xfe00000 B
GAP 8 MiB
262 MiB --- offset: 0x10600000 B --- 274726912 B --- 536576 s ---
name: logo
flags: 1
size: 8 MiB 0x800000 B
end: 270 MiB 0x10e00000 B
GAP 8 MiB
278 MiB --- offset: 0x11600000 B --- 291504128 B --- 569344 s ---
name: misc
flags: 1
size: 2 MiB 0x200000 B
end: 280 MiB 0x11800000 B
GAP 8 MiB
288 MiB --- offset: 0x12000000 B --- 301989888 B --- 589824 s ---
name: dtbo_a
flags: 1
size: 2 MiB 0x200000 B
end: 290 MiB 0x12200000 B
GAP 8 MiB
298 MiB --- offset: 0x12a00000 B --- 312475648 B --- 610304 s ---
name: dtbo_b
flags: 1
size: 2 MiB 0x200000 B
end: 300 MiB 0x12c00000 B
GAP 8 MiB
308 MiB --- offset: 0x13400000 B --- 322961408 B --- 630784 s ---
name: cri_data
flags: 2
size: 8 MiB 0x800000 B
end: 316 MiB 0x13c00000 B
GAP 8 MiB
324 MiB --- offset: 0x14400000 B --- 339738624 B --- 663552 s ---
name: param
flags: 2
size: 16 MiB 0x1000000 B
end: 340 MiB 0x15400000 B
GAP 8 MiB
348 MiB --- offset: 0x15c00000 B --- 364904448 B --- 712704 s ---
name: odm_ext_a
flags: 1
size: 16 MiB 0x1000000 B
end: 364 MiB 0x16c00000 B
GAP 8 MiB
372 MiB --- offset: 0x17400000 B --- 390070272 B --- 761856 s ---
name: odm_ext_b
flags: 1
size: 16 MiB 0x1000000 B
end: 388 MiB 0x18400000 B
GAP 8 MiB
396 MiB --- offset: 0x18c00000 B --- 415236096 B --- 811008 s ---
name: oem_a
flags: 1
size: 32 MiB 0x2000000 B
end: 428 MiB 0x1ac00000 B
GAP 8 MiB
436 MiB --- offset: 0x1b400000 B --- 457179136 B --- 892928 s ---
name: oem_b
flags: 1
size: 32 MiB 0x2000000 B
end: 468 MiB 0x1d400000 B
GAP 8 MiB
476 MiB --- offset: 0x1dc00000 B --- 499122176 B --- 974848 s ---
name: boot_a
flags: 1
size: 64 MiB 0x4000000 B
end: 540 MiB 0x21c00000 B
GAP 8 MiB
548 MiB --- offset: 0x22400000 B --- 574619648 B --- 1122304 s ---
name: boot_b
flags: 1
size: 64 MiB 0x4000000 B
end: 612 MiB 0x26400000 B
GAP 8 MiB
620 MiB --- offset: 0x26c00000 B --- 650117120 B --- 1269760 s ---
name: rsv
flags: 1
size: 16 MiB 0x1000000 B
end: 636 MiB 0x27c00000 B
GAP 8 MiB
644 MiB --- offset: 0x28400000 B --- 675282944 B --- 1318912 s ---
name: metadata
flags: 1
size: 16 MiB 0x1000000 B
end: 660 MiB 0x29400000 B
GAP 8 MiB
668 MiB --- offset: 0x29c00000 B --- 700448768 B --- 1368064 s ---
name: vbmeta_a
flags: 1
size: 2 MiB 0x200000 B
end: 670 MiB 0x29e00000 B
GAP 8 MiB
678 MiB --- offset: 0x2a600000 B --- 710934528 B --- 1388544 s ---
name: vbmeta_b
flags: 1
size: 2 MiB 0x200000 B
end: 680 MiB 0x2a800000 B
GAP 8 MiB
688 MiB --- offset: 0x2b000000 B --- 721420288 B --- 1409024 s ---
name: vbmeta_system_a
flags: 1
size: 2 MiB 0x200000 B
end: 690 MiB 0x2b200000 B
GAP 8 MiB
698 MiB --- offset: 0x2ba00000 B --- 731906048 B --- 1429504 s ---
name: vbmeta_system_b
flags: 1
size: 2 MiB 0x200000 B
end: 700 MiB 0x2bc00000 B
GAP 8 MiB
708 MiB --- offset: 0x2c400000 B --- 742391808 B --- 1449984 s ---
name: super
flags: 1
size: 2048 MiB 0x80000000 B
end: 2756 MiB 0xac400000 B
GAP 8 MiB
2764 MiB --- offset: 0xacc00000 B --- 2898264064 B --- 5660672 s ---
name: userdata
flags: 4
size: 12264 MiB 0x2fe800000 B
end: 15028 MiB 0x3ab400000 B

`

**Comment by devmfc** [comment]
This is NOT  an Armbian image.

**Comment by dmotkan** [comment]
> This is NOT an Armbian image.

Yes, this is a slimbox for Android. The Armbian recording script on emmsi gives errors when trying to record. Multiboot installation occurs without errors.

`tvbox:~#./aml-install-to-emmc.sh
Current Host disk:  [ /dev/mmcblk0 ]
Target eMMC disk: [ /dev/mmcblk1 ]
Select partition table type [ MSDOS ]
This script will erase all existing partitions and its data on the disk!!
Are you sure y/N ? y
Ok, proceeding...
Start create partition table and filesystems on eMMC [ /dev/mmcblk1 ]...
Backup vendor u-boot
Found: 27
offset: 742391808
size: 2147483648
dev             nr      name    offset          size    end             flags
Start restoring vendor u-boot
dd: error writing '/dev/mmcblk1': Input/output error
Error occured [1], create_partition_table:127, exitting
`
Interesting point. Re-flashed for slimbox TV box does not boot from the armbian SD card in the TV box's card reader without pressing the reset button. But, if you boot once from the same SD card, via a USB card reader, and then insert the SD card into the card reader of the TV box, then the Armbian is loaded  without a reset. That is, some kind of configuration is registered in the u-boot.

> Try to flash the sd card again and try again.

Yes, apparently it crashed on the first boot. Thank you

**Reply by devmfc** [reply]
I think you misunderstood. What I meant was that my images are **NOT** Armbian. 
They are plain Debian or Ubuntu.

Furthermore, my images assume there is a stock bootloader present. I don't know which bootloader Slimbox uses.

**Comment by dmotkan** [comment]
Thank you. I'll try to roll back to the factory firmware and repeat the steps. On slimbox firmware, the bootloader is guaranteed to be unlocked.

Upd.  

Unfortunately, with the factory firmware things are even worse. The TV box only boots after a reset. Emmc is not in the list of /dev . Scripts not work. Most likely the bootloader is locked. The bad thing is that the number of recording cycles for external SD card is limited.
Is it possible to somehow get feathers on emmc at least with a root partition ?

tvbox:~#./aml-part.sh
od: /dev/mmcblk1: No such file or directory
./aml-part.sh: line 76: table[0]: unbound variable

tvbox:~#./aml-multiboot-setup.sh
Configuration file wrong or corrupted
[Error]
fw_printenv has no valid config file

tvbox:~#./aml-install-to-emmc.sh
No eMMC device found!
Error occured [1], line 201, exitting

**Comment by dmotkan** [comment]
It's not a locked bootloader issue. Coreelec is flashed in emmc instead of factory firmware.

Coreelec on emmc . Boot from Ubuntu on SD-card.
Emms is not mounted.

tvbox:~#dmesg
[ 0.279927] EMMC_1V8: Failed to create debugfs directory
[    0.307141] Waiting for root device /dev/mmcblk0p2...
[    0.345857] mmc0: new high speed SDHC card at address 0001
[    0.346926] mmcblk0: mmc0:0001 SD 7.44 GiB
[    0.349646] mmcblk0: p1 p2
[    0.378315] mmc1: error -110 whilst initialising MMC card
[    0.379679] mmc1: Failed to initialize a non-removable card
[    0.395078] mmc2: new high speed SDIO card at address 0001
[    1.371372] EXT4-fs (mmcblk0p2): recovery complete
[    1.377114] EXT4-fs (mmcblk0p2): mounted filesystem 09f66b14-fa34-49d1-a1d1-528d59e87bb8 r/w with ordered data mode. Quota mode: none.
[    1.377285] VFS: Mounted root (ext4 filesystem) on device 179:2.
[    1.391445] devtmpfs: mounted

tvbox:~#vendev
wifi SDIO ID: 0x02d0:0x4330

tvbox:~#./wifi-connect.sh
Unknown device "/sys/class/net/wlan0": No such device
Connect to wifi network wlan0:
no wlan interface found

**Comment by fachos** [comment]
Hi...

i have box s905x chipset with ap6330... what i do is..

dmesg | grep brcm

```
[    3.982858] brcmfmac: brcmf_fw_alloc_request: using brcm/brcmfmac4330-sdio for chip BCM4330/4
[    3.989026] brcmfmac mmc2:0001:1: Direct firmware load for brcm/brcmfmac4330-sdio.amlogic,p212.bin failed with error -2
[    3.989968] usbcore: registered new interface driver brcmfmac
[    3.998342] brcmfmac mmc2:0001:1: Direct firmware load for brcm/brcmfmac4330-sdio.clm_blob failed with error -2
[    4.335104] brcmfmac: brcmf_c_process_clm_blob: no clm_blob available (err=-2), device may have limited channels available
[    4.340212] brcmfmac: brcmf_c_process_txcap_blob: no txcap_blob available (err=-2)
[    4.347379] brcmfmac: brcmf_c_preinit_dcmds: Firmware: BCM4330/4 wl0: Oct 25 2011 19:34:12 version 5.90.125.104
[    4.685038] ieee80211 phy0: brcmf_construct_chaninfo: Ignoring unexpected firmware channel 184
[    4.691817] ieee80211 phy0: brcmf_construct_chaninfo: Ignoring unexpected firmware channel 188
[    4.698554] ieee80211 phy0: brcmf_construct_chaninfo: Ignoring unexpected firmware channel 192
[    4.703874] ieee80211 phy0: brcmf_construct_chaninfo: Ignoring unexpected firmware channel 196
[    4.711091] ieee80211 phy0: brcmf_construct_chaninfo: Ignoring unexpected firmware channel 200
[    4.711101] ieee80211 phy0: brcmf_construct_chaninfo: Ignoring unexpected firmware channel 204
[    4.711106] ieee80211 phy0: brcmf_construct_chaninfo: Ignoring unexpected firmware channel 208
[    4.725450] ieee80211 phy0: brcmf_construct_chaninfo: Ignoring unexpected firmware channel 212
[    4.725464] ieee80211 phy0: brcmf_construct_chaninfo: Ignoring unexpected firmware channel 216
```
found out ap6330 using brcmfmac4330

cd /lib/firmware/brcm/
ls

```
BCM4330B1.hcd  BCM4335C0.hcd  BCM4356A2.hcd                    brcmfmac4330-sdio.txt   brcmfmac4335-sdio.bin  brcmfmac4339-sdio.txt        brcmfmac43455-sdio.txt      brcmfmac4354-sdio.txt       brcmfmac4356-sdio.txt
BCM43342.hcd   BCM4345C0.hcd  brcmfmac4330-sdio.bin            brcmfmac43342-sdio.bin  brcmfmac4335-sdio.txt  brcmfmac43455-sdio.bin       brcmfmac4354-sdio.bin       brcmfmac4356-sdio.bin       nvram_ap6354.txt
BCM4335A0.hcd  BCM4354.hcd    brcmfmac4330-sdio,tanix-tx3.txt  brcmfmac43342-sdio.txt  brcmfmac4339-sdio.bin  brcmfmac43455-sdio.clm_blob  brcmfmac4354-sdio.clm_blob  brcmfmac4356-sdio.clm_blob
```

here 3 files with *4330 :- 

**brcmfmac4330-sdio.bin
brcmfmac4330-sdio.txt
brcmfmac4330-sdio,tanix-tx3.txt**

i just copy 2 files to :-

**cp brcmfmac4330-sdio.bin brcmfmac4330-sdio.amlogic,p212.bin
cp brcmfmac4330-sdio.txt brcmfmac4330-sdio.amlogic,p212.txt**

reboot....

./wifi-connect.sh

```
Connect to wifi network wlan0: brcmfmac

1: kelvin109@unifi
2: MYWIFI123
3: simple-family
Choose wifi SSID: 2
Please enter password/pre shared key for [MYWIFI123]:
Trying to authenticate, please wait...
auth ok
Saving wifi config...
Try getting ip from DHCP...

Ready. Wifi address: 192.168.0.140
```

now im online using AP6330..  STILL error show in dmesg.... 

```
[    3.981992] brcmfmac mmc2:0001:1: Direct firmware load for brcm/brcmfmac4330-sdio.clm_blob failed with error -2
[    4.296848] brcmfmac: brcmf_c_process_clm_blob: no clm_blob available (err=-2), device may have limited channels available
[    4.302124] brcmfmac: brcmf_c_process_txcap_blob: no txcap_blob available (err=-2)
[    4.308711] brcmfmac: brcmf_c_preinit_dcmds: Firmware: BCM4330/4 wl0: Oct 25 2011 19:34:12 version 5.90.125.104
[    4.616714] ieee80211 phy0: brcmf_construct_chaninfo: Ignoring unexpected firmware channel 184
[    4.621566] ieee80211 phy0: brcmf_construct_chaninfo: Ignoring unexpected firmware channel 188
[    4.626439] ieee80211 phy0: brcmf_construct_chaninfo: Ignoring unexpected firmware channel 192
[    4.631423] ieee80211 phy0: brcmf_construct_chaninfo: Ignoring unexpected firmware channel 196
[    4.636326] ieee80211 phy0: brcmf_construct_chaninfo: Ignoring unexpected firmware channel 200
[    4.641331] ieee80211 phy0: brcmf_construct_chaninfo: Ignoring unexpected firmware channel 204
[    4.646303] ieee80211 phy0: brcmf_construct_chaninfo: Ignoring unexpected firmware channel 208
[    4.651347] ieee80211 phy0: brcmf_construct_chaninfo: Ignoring unexpected firmware channel 212
[    4.658531] ieee80211 phy0: brcmf_construct_chaninfo: Ignoring unexpected firmware channel 216
```

brcmfmac4330 doesnt come with clm_blob file.... but its ok for me

you can try but change amlogic,p212 because im using s905x chipset...
check your dmesg

hope this will solve your wifi problem

**Reply by Adney** [reply]
I tried it on T95W2 even before your post appeared, but wi-fi/bt did not work.

brcmfmac4330-sdio.Tanix W2.bin
brcmfmac4330-sdio.Tanix W2.txt

**Reply by fachos** [reply]
Impossible Tanix(space)W2
Tanix,W2

check dmesg output for the brcm

**Reply by Adney** [reply]
**After renaming**

[ 6.044672] brcmfmac: brcmf_fw_alloc_request: using brcm/brcmfmac4330-sdio for chip BCM4330/4
[ 6.050322] usbcore: registered new interface driver brcmfmac
[ 6.123218] brcmfmac mmc2:0001:1: Direct firmware load for brcm/brcmfmac4330-sdio.clm_blob failed with error -2

**Comment by fachos** [comment]
Hi

Show me dmesg | grep brcm

renaming should same what dmesg output

**Reply by Adney** [reply]
Hi,
**Before the renaming, it was like this**

[ 6.053215] brcmfmac: brcmf_fw_alloc_request: using brcm/brcmfmac4330-sdio for chip BCM4330/4
[ 6.056699] usbcore: registered new interface driver brcmfmac
[ 6.065308] brcmfmac mmc2:0001:1: Direct firmware load for brcm/brcmfmac4330-sdio.Tanix W2.bin failed with error -2
[ 6.099428] brcmfmac mmc2:0001:1: Direct firmware load for brcm/brcmfmac4330-sdio.clm_blob failed with error -2
